haystack/preview/components/fetchers/link_content.py

```diff
@@ -0,0 +1,225 @@
+from haystack.preview import Document
+
+logger = logging.getLogger(__name__)
+
+
+def text_content_handler(response: Response) -> Optional[str]:
+    """
+    :param response: Response object from the request.
+    :return: The extracted text.
+    """
+    return response.text
+
+
+def binary_content_handler(response: Response) -> IO[bytes]:
+    """
+    :param response: Response object from the request.
+    :return: The extracted binary file-like object.
+    """
+    return io.BytesIO(response.content)
+
+
+@component
+class LinkContentFetcher:
+    """
+    LinkContentFetcher fetches content from a URL link
+    """
+
+    outgoing_edges = 1
+
+    _USER_AGENT = f"haystack/LinkContentRetriever/{__version__}"
+
+    _REQUEST_HEADERS = {
+        "accept": "*/*",
+        "User-Agent": _USER_AGENT,
+        "Accept-Language": "en-US,en;q=0.9,it;q=0.8,es;q=0.7",
+        "referer": "https://www.google.com/",
+    }
+
+    def __init__(
+        self,
+        raise_on_failure: Optional[bool] = False,
+        user_agents: Optional[List[str]] = None,
+        retry_attempts: Optional[int] = None,
+    ):
+        """
+
+        Creates a LinkContentFetcher instance.
+        :param raise_on_failure: A boolean indicating whether to raise an exception when a failure occurs
+                         during content extraction. If False, the error is simply logged and the program continues.
+                         Defaults to False.
+        :param user_agents: A list of user agents to use when fetching content. Defaults to None.
+        :param retry_attempts: The number of times to retry fetching content. Defaults to 2.
+        """
+        super().__init__()
+        self.raise_on_failure = raise_on_failure
+        self.user_agents = user_agents or [LinkContentFetcher._USER_AGENT]
+        self.current_user_agent_idx: int = 0
+        self.retry_attempts = retry_attempts or 2
+        self.handlers: Dict[str, Callable] = defaultdict(lambda: text_content_handler)
+
+        # register default content handlers that extract data from the response
+        self._register_content_handler("text/html", text_content_handler)
+        self._register_content_handler("text/plain", text_content_handler)
+        self._register_content_handler("application/pdf", binary_content_handler)
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Serialize this component to a dictionary.
+        """
+        return default_to_dict(
+            self,
+            raise_on_failure=self.raise_on_failure,
+            user_agents=self.user_agents,
+            retry_attempts=self.retry_attempts,
+        )
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "LinkContentFetcher":
+        """
+        Deserialize this component from a dictionary.
+        """
+        return default_from_dict(cls, data)
+
+    def fetch(self, url: str, timeout: Optional[int] = 3, doc_kwargs: Optional[dict] = None) -> Document:
```

Comment:
Let's see if we really need it because it really complicates things (error handling for each request/response, returned docs). That was all previously done in WebRetriever